lib/events/filesessions/fileasync_test.go

```diff
@@ -359,7 +359,199 @@ func TestUploadResume(t *testing.T) {
+	stream, err := fileStreamer.CreateAuditStream(ctx, session.ID(sid))
+	assert.NoError(t, err)
+
+	for _, event := range inEvents {
+		err := stream.EmitAuditEvent(ctx, event)
+		assert.NoError(t, err)
+	}
+	err = stream.Complete(ctx)
+	assert.NoError(t, err)
+
+	// initiate the scan by advancing clock past
+	// block period
+	clock.Advance(scanPeriod + time.Second)
+
+	// initiate several scan attempts and increase the scan period
+	attempts := 10
+	var prev time.Time
+	var diffs []time.Duration
+	for i := 0; i < attempts; i++ {
+		// wait for the upload event
+		var event events.UploadEvent
+		select {
+		case event = <-eventsC:
+			assert.Error(t, event.Error)
+			if prev.IsZero() {
+				prev = event.Created
+			} else {
+				diffs = append(diffs, event.Created.Sub(prev))
+				prev = event.Created
+			}
+		case <-ctx.Done():
+			t.Fatalf("Timeout waiting for async upload %v, try `go test -v` to get more logs for details", i)
+		}
+
+		// After in Scan has been called two times,
+		// first time after doing the scan, and second
+		// on receiving the event to <- eventsCh
+		clock.BlockUntil(2)
+		clock.Advance(scanPeriod*time.Duration(i+2) + time.Second)
+	}
+
+	// Make sure that durations between retries are increasing
+	for i, diff := range diffs {
+		if i > 0 {
+			assert.Equal(t, true, diff > diffs[i-1], "Expected next retry to take longer, got %v vs %v", diffs[i-1], diff)
+		}
+	}
+
+	// Fix the streamer, make sure the upload succeeds
+	terminateConnectionAt.Store(0)
+	clock.BlockUntil(2)
+	clock.Advance(time.Hour)
+	select {
+	case event := <-eventsC:
+		assert.NoError(t, event.Error)
+		if prev.IsZero() {
+			prev = event.Created
+		} else {
+			prev = event.Created
+		}
+	case <-ctx.Done():
+		t.Fatalf("Timeout waiting for async upload, try `go test -v` to get more logs for details")
+	}
+}
+
+// TestUploadBadSession creates a corrupted session file
+// and makes sure the uploader marks it as faulty
+func TestUploadBadSession(t *testing.T) {
+	utils.InitLoggerForTests(testing.Verbose())
+
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	clock := clockwork.NewFakeClock()
+	eventsC := make(chan events.UploadEvent, 100)
+	memUploader := events.NewMemoryUploader(eventsC)
+	streamer, err := events.NewProtoStreamer(events.ProtoStreamerConfig{
+		Uploader:       memUploader,
+		MinUploadBytes: 1024,
+	})
+	assert.NoError(t, err)
+
+	scanDir, err := ioutil.TempDir("", "teleport-streams")
+	assert.NoError(t, err)
+	defer os.RemoveAll(scanDir)
+
+	scanPeriod := 10 * time.Second
+	uploader, err := NewUploader(UploaderConfig{
+		Context:    ctx,
+		ScanDir:    scanDir,
+		ScanPeriod: scanPeriod,
+		Streamer:   streamer,
+		Clock:      clock,
+		EventsC:    eventsC,
+	})
+	assert.NoError(t, err)
+	go uploader.Serve()
+	// wait until uploader blocks on the clock
+	clock.BlockUntil(1)
+
+	defer uploader.Close()
+
+	// wait until uploader blocks on the clock before creating the stream
+	clock.BlockUntil(1)
```

Comment:
nit: can some of this startup boilerplate move to a helper func?

Reply:
good point, I will improve in 5.0 master branch